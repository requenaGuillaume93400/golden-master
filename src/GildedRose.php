<?php

declare(strict_types=1);

namespace GildedRose;

final class GildedRose
{
    /**
     * @param Item[] $items
     */
    public function __construct(
        private array $items
    ) {
    }

    // public function updateQuality(): void
    // {
    //     foreach ($this->items as $item) {
    //         if ($item->name != 'Aged Brie' and $item->name != 'Backstage passes to a TAFKAL80ETC concert') {
    //             if ($item->quality > 0) {
    //                 if ($item->name != 'Sulfuras, Hand of Ragnaros') {
    //                     $item->quality = $item->quality - 1;
    //                 }
    //             }
    //         } else {
    //             if ($item->quality < 50) {
    //                 $item->quality = $item->quality + 1;
    //                 if ($item->name == 'Backstage passes to a TAFKAL80ETC concert') {
    //                     if ($item->sellIn < 11) {
    //                         if ($item->quality < 50) {
    //                             $item->quality = $item->quality + 1;
    //                         }
    //                     }
    //                     if ($item->sellIn < 6) {
    //                         if ($item->quality < 50) {
    //                             $item->quality = $item->quality + 1;
    //                         }
    //                     }
    //                 }
    //             }
    //         }

    //         if ($item->name != 'Sulfuras, Hand of Ragnaros') {
    //             $item->sellIn = $item->sellIn - 1;
    //         }

    //         if ($item->sellIn < 0) {
    //             if ($item->name != 'Aged Brie') {
    //                 if ($item->name != 'Backstage passes to a TAFKAL80ETC concert') {
    //                     if ($item->quality > 0) {
    //                         if ($item->name != 'Sulfuras, Hand of Ragnaros') {
    //                             $item->quality = $item->quality - 1;
    //                         }
    //                     }
    //                 } else {
    //                     $item->quality = $item->quality - $item->quality;
    //                 }
    //             } else {
    //                 if ($item->quality < 50) {
    //                     $item->quality = $item->quality + 1;
    //                 }
    //             }
    //         }
    //     }
    // }

    const AGED_BRIE = 'Aged Brie';
    const BACKSTAGE = 'Backstage passes to a TAFKAL80ETC concert';
    const SULFURAS = 'Sulfuras, Hand of Ragnaros';


    public function updateQuality(): void
    {
        foreach ($this->items as $item) {
            /** @var Item $item */
            if (!in_array($item->name, [self::AGED_BRIE, self::BACKSTAGE])) {
                if ($item->quality > 0) {
                    if ($item->name != self::SULFURAS) {
                        $item->quality--;
                    }
                }
            } else {
                if ($item->quality < 50) {
                    $item->quality++;
                    if ($item->name == self::BACKSTAGE) {
                        if ($item->sellIn < 11) {
                            $this->incrementQualityIfInferiorToFifty($item);
                        }
                        if ($item->sellIn < 6) {
                            $this->incrementQualityIfInferiorToFifty($item);
                        }
                    }
                }
            }

            if ($item->name != self::SULFURAS) {
                $item->sellIn--;
            }

            // finir ici
            if ($item->sellIn < 0) {
                if ($item->name != self::AGED_BRIE) {
                    if ($item->name != self::BACKSTAGE) {
                        if ($item->quality > 0) {
                            if ($item->name != self::SULFURAS) {
                                $item->quality--;
                            }
                        }
                    } else {
                        $item->quality = 0;
                    }
                } else {
                    $this->incrementQualityIfInferiorToFifty($item);
                }
            }
        }
    }

    private function updateQualityAgedBrie(Item $item): void
    {
        $this->incrementQualityIfInferiorToFifty($item);
        $item->sellIn--;

        if($item->sellIn < 0){
            $this->incrementQualityIfInferiorToFifty($item);
        }
    }

    private function updateQualityBackstage(Item $item): void
    {
        $this->incrementQualityIfInferiorToFifty($item);

        if ($item->sellIn < 11) {
            $this->incrementQualityIfInferiorToFifty($item);
        }

        if ($item->sellIn < 6) {
            $this->incrementQualityIfInferiorToFifty($item);
        }

        $item->sellIn--;
        $item->quality = 0;
    }

    private function updateQualitySulfuras(Item $item): void
    {
        
    }

    private function updateQualityOther(Item $item): void
    {
        if ($item->quality > 0) {
            $item->quality--;
        }

        $item->sellIn--;

        if ($item->sellIn < 0 && $item->quality > 0) {
            $item->quality--;
        }
    }

    private function incrementQualityIfInferiorToFifty(Item $item): void
    {
        if ($item->quality < 50) {
            $item->quality++;
        }
    }

}