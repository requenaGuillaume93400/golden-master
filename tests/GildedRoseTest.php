<?php

declare(strict_types=1);

namespace Tests;

use ApprovalTests\Approvals;
use ApprovalTests\CombinationApprovals;
use GildedRose\GildedRose;
use GildedRose\Item;
use PHPUnit\Framework\TestCase;

class GildedRoseTest extends TestCase
{
    public function testFoo(): void
    {
        // Test sur une valeur
        // $items = [new Item('foo', 1000, 1000)];
        // $gildedRose = new GildedRose($items);
        // $gildedRose->updateQuality();
        // Approvals::verifyString($items[0]->__toString());

        // Tester tous les cas en se basant sur les valeurs qu'on a repéré dans les if
        $names = ['Aged Brie', 'Backstage passes to a TAFKAL80ETC concert', 'Sulfuras, Hand of Ragnaros', 'foo'];
        $sellIns = [5, 12, -1, 7];
        $qualities = [45, 51, -1];

        CombinationApprovals::verifyAllCombinations3(
            function (string $name, int $sell_in, int $quality){
                $items = [new Item($name, $sell_in, $quality)];
                $gildedRose = new GildedRose($items);
                $gildedRose->updateQuality();        
                return $items[0]->__toString(); // return au lieu de approval verifyString ici
            },
            $names,
            $sellIns,
            $qualities
        );
    }

    // Principes
    /**
     * 1) On observe ce que le code renvoit en jouant un 1er test,
     * le resultat sera ecrit dans le .received.txt
     * on copie colle ce resultat dans .approved.txt
     * 
     * 2) On fait bien attention a tester tous les cas grace au code coverage
     * qui doit atteindre les 100%
     * 
     * 3) A chaque test lancé, la librairie comparera les resultats dans
     * received et approved (received est généré a chaque test, approved ne bouge pas)
     * et si on a un echec, c'est qu'on a modifier le comportement du code, on a fait une erreur
     */
}