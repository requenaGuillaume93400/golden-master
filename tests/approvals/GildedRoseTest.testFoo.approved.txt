[Aged Brie, 5, 45] => Aged Brie, 4, 46
[Aged Brie, 5, 51] => Aged Brie, 4, 51
[Aged Brie, 5, -1] => Aged Brie, 4, 0
[Aged Brie, 12, 45] => Aged Brie, 11, 46
[Aged Brie, 12, 51] => Aged Brie, 11, 51
[Aged Brie, 12, -1] => Aged Brie, 11, 0
[Aged Brie, -1, 45] => Aged Brie, -2, 47
[Aged Brie, -1, 51] => Aged Brie, -2, 51
[Aged Brie, -1, -1] => Aged Brie, -2, 1
[Aged Brie, 7, 45] => Aged Brie, 6, 46
[Aged Brie, 7, 51] => Aged Brie, 6, 51
[Aged Brie, 7, -1] => Aged Brie, 6, 0
[Backstage passes to a TAFKAL80ETC concert, 5, 45] => Backstage passes to a TAFKAL80ETC concert, 4, 48
[Backstage passes to a TAFKAL80ETC concert, 5, 51] => Backstage passes to a TAFKAL80ETC concert, 4, 51
[Backstage passes to a TAFKAL80ETC concert, 5, -1] => Backstage passes to a TAFKAL80ETC concert, 4, 2
[Backstage passes to a TAFKAL80ETC concert, 12, 45] => Backstage passes to a TAFKAL80ETC concert, 11, 46
[Backstage passes to a TAFKAL80ETC concert, 12, 51] => Backstage passes to a TAFKAL80ETC concert, 11, 51
[Backstage passes to a TAFKAL80ETC concert, 12, -1] => Backstage passes to a TAFKAL80ETC concert, 11, 0
[Backstage passes to a TAFKAL80ETC concert, -1, 45] => Backstage passes to a TAFKAL80ETC concert, -2, 0
[Backstage passes to a TAFKAL80ETC concert, -1, 51] => Backstage passes to a TAFKAL80ETC concert, -2, 0
[Backstage passes to a TAFKAL80ETC concert, -1, -1] => Backstage passes to a TAFKAL80ETC concert, -2, 0
[Backstage passes to a TAFKAL80ETC concert, 7, 45] => Backstage passes to a TAFKAL80ETC concert, 6, 47
[Backstage passes to a TAFKAL80ETC concert, 7, 51] => Backstage passes to a TAFKAL80ETC concert, 6, 51
[Backstage passes to a TAFKAL80ETC concert, 7, -1] => Backstage passes to a TAFKAL80ETC concert, 6, 1
[Sulfuras, Hand of Ragnaros, 5, 45] => Sulfuras, Hand of Ragnaros, 5, 45
[Sulfuras, Hand of Ragnaros, 5, 51] => Sulfuras, Hand of Ragnaros, 5, 51
[Sulfuras, Hand of Ragnaros, 5, -1] => Sulfuras, Hand of Ragnaros, 5, -1
[Sulfuras, Hand of Ragnaros, 12, 45] => Sulfuras, Hand of Ragnaros, 12, 45
[Sulfuras, Hand of Ragnaros, 12, 51] => Sulfuras, Hand of Ragnaros, 12, 51
[Sulfuras, Hand of Ragnaros, 12, -1] => Sulfuras, Hand of Ragnaros, 12, -1
[Sulfuras, Hand of Ragnaros, -1, 45] => Sulfuras, Hand of Ragnaros, -1, 45
[Sulfuras, Hand of Ragnaros, -1, 51] => Sulfuras, Hand of Ragnaros, -1, 51
[Sulfuras, Hand of Ragnaros, -1, -1] => Sulfuras, Hand of Ragnaros, -1, -1
[Sulfuras, Hand of Ragnaros, 7, 45] => Sulfuras, Hand of Ragnaros, 7, 45
[Sulfuras, Hand of Ragnaros, 7, 51] => Sulfuras, Hand of Ragnaros, 7, 51
[Sulfuras, Hand of Ragnaros, 7, -1] => Sulfuras, Hand of Ragnaros, 7, -1
[foo, 5, 45] => foo, 4, 44
[foo, 5, 51] => foo, 4, 50
[foo, 5, -1] => foo, 4, -1
[foo, 12, 45] => foo, 11, 44
[foo, 12, 51] => foo, 11, 50
[foo, 12, -1] => foo, 11, -1
[foo, -1, 45] => foo, -2, 43
[foo, -1, 51] => foo, -2, 49
[foo, -1, -1] => foo, -2, -1
[foo, 7, 45] => foo, 6, 44
[foo, 7, 51] => foo, 6, 50
[foo, 7, -1] => foo, 6, -1
